parser-typechecker/src/Unison/CommandLine.hs

```diff
@@ -56,20 +61,47 @@ watchFileSystem q dir = do
     atomically . Q.enqueue q $ UnisonFileChanged (Text.pack filePath) text
   pure (cancel >> killThread t)
 
-watchBranchUpdates :: IO Branch.Hash -> TQueue Event -> Codebase IO v a -> IO (IO ())
+watchBranchUpdates :: IO (Branch.Branch IO) -> TQueue Event -> Codebase IO v a -> IO (IO ())
 watchBranchUpdates currentRoot q codebase = do
   (cancelExternalBranchUpdates, externalBranchUpdates) <-
     Codebase.rootBranchUpdates codebase
   thread <- forkIO . forever $ do
     updatedBranches <- externalBranchUpdates
     currentRoot <- currentRoot
-    -- We only issue the event if the branch is different than what's already
-    -- in memory. This skips over file events triggered by saving to disk what's
-    -- already in memory.
-    when (any (/= currentRoot) updatedBranches) $
-      atomically . Q.enqueue q . IncomingRootBranch $ Set.delete currentRoot updatedBranches
+    -- Since there's some lag between when branch files are written and when
+    -- the OS generates a file watch event, we skip branch update events
+    -- that are causally before the current root.
+    --
+    -- NB: Sadly, since the file watching API doesn't have a way to silence
+    -- the events from a specific individual write, this is ultimately a
+    -- heuristic. If a fairly recent head gets deposited at just the right
+    -- time, it would get ignored by this logic. This seems unavoidable.
```

Comment:
I may be being dense, but it seems to me that given that events get generated for files written to disk, and we exactly know which ones we are writing (because we are writing them) that we could exactly silence exactly those.

e.g. When we write a root, we add it to a set; when we receive an event, we remove it from the set.  If we receive an event for something that isn't in the set of the things we said we wrote, then we didn't write that.

Right? 😅  And then wouldn't have to wait for a `IO (Branch IO)` to load, and for a Branch.before to run (or truncate).

Reply:
That relies for correctness on the file watching not missing any writes. Otherwise this set you maintain could accumulate stale hashes in it. I didn't like that approach, just seemed fragile to me, also more complicated to implement and test I thought (though I could be wrong about that).